src/matrix/storage/idb/Store.ts

```diff
@@ -0,0 +1,175 @@
+/*
+Copyright 2020 Bruno Windels <bruno@windels.cloud>
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import {QueryTarget, IDBQuery} from "./QueryTarget";
+import {IDBRequestAttemptError} from "./error";
+import {reqAsPromise} from "./utils";
+import {Transaction} from "./Transaction";
+
+const LOG_REQUESTS = false;
+
+function logRequest(method: string, params: any[], source: any): void {
+    const storeName = source?.name;
+    const databaseName = source?.transaction?.db?.name;
+    console.info(`${databaseName}.${storeName}.${method}(${params.map(p => JSON.stringify(p)).join(", ")})`);
+}
+
+class QueryTargetWrapper<T> {
+    private _qt: IDBIndex | IDBObjectStore;
+
+    constructor(qt: IDBIndex | IDBObjectStore) {
+        this._qt = qt;
+    }
+
+    get keyPath(): string | string[] {
+        if (this._qt["objectStore"]) {
+            return (this._qt as IDBIndex).objectStore.keyPath;
+        } else {
+            return this._qt.keyPath;
+        }
+    }
+
+    get _qtStore(): IDBObjectStore {
+        return this._qt as IDBObjectStore;
```

Comment:
I doubt we use `add` or `put` on an IDBIndex currently, but perhaps we can actually do the right thing here like so:
```ts
get _qtStore(): IDBObjectStore {
    if ("objectStore" in this._qt) {
        return this._qt.objectStore;
    }
    return this._qt;
}
```

Reply:
Is that equivalent to the current behavior? I _think_ that currently the code just tries to call add/subtract and throws an exception if not, so unsafe casting seemed to be a similar approach in spirit. But if we don't use `add` or `put` on an `IDBIndex` (if we did, we'd be seeing exceptions :smile: ) , then it's probably safe to change the behavior of these functions.

Reply:
Seems safe to me indeed